src/main/java/frc/robot/Constants.java

```diff
@@ -8,7 +8,131 @@
+        public static final double GEAR_RATIO_DRIVE_MOTOR = 7.5;
+        public static final double GEAR_RATIO_ANGLE_MOTOR = 1;
+        public static final double DRIVE_MOTOR_TICKS_PER_METER = GEAR_RATIO_DRIVE_MOTOR * TICKS_PER_ROTATION_DRIVE_MOTOR / (4 * 0.0254 * Math.PI);
+        public static final double ANGLE_MOTOR_TICKS_PER_RADIAN = GEAR_RATIO_ANGLE_MOTOR * TICKS_PER_ROTATION_ANGLE_MOTOR / (2 * Math.PI);
+
+        public static final int MAX_CURRENT = 15; // [amps]
+
+        // State Space
+        public static final double VELOCITY_TOLERANCE = 5; // [rps]
+        public static final double COST_LQR = 11;
+        // Note that the values of MODEL_TOLERANCE and ENCODER_TOLERANCE should be a lot smaller (something like 1e-6)
+        public static final double MODEL_TOLERANCE = 0.01;
+        public static final double ENCODER_TOLERANCE = 0.01; // [ticks]
+
+        public static final double THETA_KP = 1;
+        public static final double THETA_KI = 1;
+        public static final double THETA_KD = 1;
+
+        // The theta is responsible for the angle of the whole chassis, while the angle is used in the angle motor itself.
+        public static final double ALLOWABLE_THETA_ERROR = Math.toRadians(0.05); // [rad]
+        public static final double ALLOWABLE_ANGLE_ERROR = Math.toRadians(8); // [rad]
+        public static final double WHEEL_RADIUS = 0.04688; // [m]
+
+        public static final double ROBOT_LENGTH = 0.5924; // [m]
+        public static final double ROBOT_WIDTH = 0.5924; // [m]
+        // the rotational velocity of the robot, this constant multiplies the rotation output of the joystick
+        public static final double JOYSTICK_THRESHOLD = 0.1; // [%]
+        public static final double OUTER_JOYSTICK_THRESHOLD = 0.95; // [%]
+        public static final double JOYSTICK_ANGLE_DEADZONE = 5; // [degrees]
+        public static final double ROTATION_DELAY = 0.1; // [sec]
+        public static final int ANGLE_CURVE_STRENGTH = 4;
+        private static final double Rx = SwerveDrive.ROBOT_WIDTH / 2;
+        private static final double Ry = SwerveDrive.ROBOT_LENGTH / 2;
+        // angle motion magic
+        private static final float MOTION_MAGIC_SAFETY = 0.7f;
+        public static final int ANGLE_MOTION_ACCELERATION = Math.round(2800 * MOTION_MAGIC_SAFETY);
+        public static final int ANGLE_CRUISE_VELOCITY = Math.round(550 * MOTION_MAGIC_SAFETY);
+        // Axis systems
+        // TODO: check the coordinate system.
+        private static final double[] signX = {1, 1, -1, -1};
+        private static final double[] signY = {-1, 1, -1, 1};
+        public static final Translation2d[] SWERVE_POSITIONS = new Translation2d[]{
+                new Translation2d(signX[0] * Rx, signY[0] * Ry),
+                new Translation2d(signX[1] * Rx, signY[1] * Ry),
+                new Translation2d(signX[2] * Rx, signY[2] * Ry),
+                new Translation2d(signX[3] * Rx, signY[3] * Ry)
```

Comment:
I think it's clearer to just set the inversions directly, this saves looking at the arrays.
Also, the declerations of `Rx` and `Ry` should move here and not separated.
```suggestion
        public static final Translation2d[] SWERVE_POSITIONS = new Translation2d[]{
                new Translation2d(Rx, -Ry),
                new Translation2d(Rx, Ry),
                new Translation2d(-Rx, -Ry),
                new Translation2d(-Rx, Ry)
```